File: 70-climbing-stairs/climbing-stairs.js
/**
This problem is a variation of the Fibonacci sequence, where the number of ways to reach step n is the sum of the ways to reach steps n-1 and n-2.
 */
var climbStairs = function (n) {
    if (n <= 2) return n;

    let prev1 = 2;
    let prev2 = 1;
    let curr;

    for (let i = 3; i <= n; i++) {
        curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }

    return prev1;
};

File: 70-climbing-stairs/climbing-stairs.cpp
const auto _ = std::cin.tie(nullptr)->sync_with_stdio(false);

class Solution {
private:
    int solve(int steps, vector<int>& dp) {
        if (steps == 0) {
            return 1;
        }
        if (steps < 0) {
            return 0;
        }

        if (dp[steps] != -1) {
            return dp[steps];
        }
        return dp[steps] = solve(steps - 1, dp) + solve(steps - 2, dp);
    }

public:
    int climbStairs(int n) {
        vector<int> dp(n + 1, 0);
        dp[0] = dp[1] = 1;

        for (int step = 2; step <= n; step++) {
            dp[step] = dp[step - 1] + dp[step - 2];
        }
        return dp[n];
    }
};
